### PR TITLE
Revamp the style guide formatting, rewrite it, and add a few rules.

### PR DESCRIPTION
**Pull request by csilvers**
Summary:
Maybe I should have done this as separate commits, but the file is
small enough I think it's reasonable to deal with them all in one go.

Major formatting changes:

* We use h3 and h4 rather than h2 and h3, which makes more of the
  style guide fit on a screen at a time, and in general improves
  readability (I think).
* The style guide is now divided into logical sections ("formatting,"
  "libraries," etc.), each with a small number of items in them.  I
  think this improves readability as well.
* I have standardized on 'yes'/'no' text for examples, rather than
  sometimes having 'good'/'bad'.  I like 'yes' or 'no': style guides
  should be opinionated.

Likewise, I reworded a bunch of stuff to say "do this, do that" rather
than just "this is good, that is good."  Other rewording was mostly to
tighten the text.  I also got rid of some obsolete text (such as a
mention of the new package system, which never materialized).  I also
tried to rewrite the 'standard components' section so it doesn't get
out of date quite so quickly.

I added two new rules:

* Only export a single react class.  This came out of the best
  practices doc at
  https://docs.google.com/document/d/1ChtFUao18IyNhaXZ5sE2W-CFuFcYnqlFTyi5gfe6XV0/edit#heading=h.10yidnt3j9ml
* Some discussion about "dumb" and "smart" components.

I also tightened the rule about using propTypes.

Test Plan:
I installed a markdown viewer plugin in my browser and visited
   file:///home/csilvers/khan/style-guides/style/react.md
and saw that the page rendered in a way that looked good, at least to
me.

### CODE REVIEW

File: style/react.md [hunk 1]
```diff
@@ -74,29 +39,128 @@ return (
...
+because what's displayed no longer directly tracks what's stored.  By
+having the data in one place instead of copying it, you ensure
+everything is consistent.
+
+#### Use [propTypes](http://facebook.github.io/react/docs/reusable-components.html).
+
+React Components should always have complete propTypes.  Every
+attribute of this.props should have a corresponding entry in
+propTypes.  This documents that props need to be passed to a model.
+([example](http://jsfiddle.net/spicyj/DEpwb/))
```
Comment (xymostech) [inline style/react.md]: Maybe link to an example from our own code? https://github.com/Khan/webapp/blob/32aa862769d4e93c477dc0ee0388816056252c4a/javascript/search-package/search-results-list.jsx#L14

File: style/react.md [hunk 2]
```diff
@@ -74,29 +39,128 @@ return (
...
+themselves.
+
+"Dumb" components are typically reusable, and do emit HTML.
+
+Smart components can have internal state, but dumb components never
+should.
+
+#### *Never* store state in the DOM.
+
+Do not use data- attributes or classes).  All information should be
```
Comment (xymostech) [inline style/react.md]: nit: extra ), and maybe code-ify `data-`

**Comment by csilvers** [comment]
Will re-do the pull request from another branch to make it easier to manage.

**Comment by rileyjshaw** [comment]
Whoops - @csilvers it looks like I commented on the wrong PR, let me know if you want me to move these to #24

**Pull request closed without merging**